packages/hardhat-core/src/internal/hardhat-network/provider/provider.ts

```diff
@@ -393,3 +393,82 @@ export class HardhatNetworkProvider
     this.emit("message", message);
   }
 }
+
+class EdrProviderWrapper extends EventEmitter implements EIP1193Provider {
+  private _provider: EdrProviderT;
+
+  public static async create(
+    config: HardhatNetworkProviderConfig
+  ): Promise<EdrProviderWrapper> {
+    const { Provider } =
+      require("@ignored/edr") as typeof import("@ignored/edr");
+
+    const coinbase = config.coinbase ?? DEFAULT_COINBASE;
+
+    const provider = await Provider.withConfig({
+      chainId: BigInt(config.chainId),
+      cacheDir: config.forkCachePath,
+      coinbase: Buffer.from(coinbase.slice(2), "hex"),
+      hardfork: ethereumsjsHardforkToEdrSpecId(
+        getHardforkName(config.hardfork)
+      ),
+      networkId: BigInt(config.chainId),
+      blockGasLimit: BigInt(config.blockGasLimit),
+      genesisAccounts: config.genesisAccounts.map((account) => {
+        return {
+          secretKey: account.privateKey,
+          balance: BigInt(account.balance),
+        };
+      }),
+      allowUnlimitedContractSize: config.allowUnlimitedContractSize,
+      allowBlocksWithSameTimestamp:
+        config.allowBlocksWithSameTimestamp ?? false,
+      initialBaseFeePerGas: BigInt(
+        config.initialBaseFeePerGas ?? 1_000_000_000
+      ),
+    });
+
+    return new EdrProviderWrapper(provider);
+  }
+
+  private constructor(provider: EdrProviderT) {
+    super();
+
+    this._provider = provider;
+  }
+
+  public async request(args: RequestArguments): Promise<unknown> {
+    const stringifiedArgs = JSON.stringify(args);
+    const response = await this._provider.handleRequest(stringifiedArgs);
+    return JSON.parse(response);
+  }
+}
+
+export async function createHardhatNetworkProvider(
+  hardhatNetworkProviderConfig: HardhatNetworkProviderConfig,
+  logger: ModulesLogger,
+  artifacts?: Artifacts
+): Promise<EIP1193Provider> {
+  let eip1193Provider: EIP1193Provider;
+
+  const vmModeEnvVar = process.env.HARDHAT_EXPERIMENTAL_VM_MODE ?? "ethereumjs";
+
+  if (vmModeEnvVar === "edr") {
+    eip1193Provider = await EdrProviderWrapper.create(
+      hardhatNetworkProviderConfig
+    );
+  } else if (vmModeEnvVar === "ethereumjs") {
```

Comment:
I think we should also add the dual mode adapter to use `HardhatNetworkProvider`. That way we'll be running the `DualModeAdapter` at the component-level still, maintaining testing support for any changes that happen at that level.